Added SelectAll SelectNone feature in CiviCRM static options element

aaayyuusshh opened the pull request:
Drupal Issue URL: https://www.drupal.org/project/webform_civicrm/issues/3224544

Work in progress: Checking off this checkbox will not enable all countries(check off checkboxes below it) like it is supposed to yet.

Overview
----------------------------------------
Added a checkbox beside "ENABLED"  in the static options for the country field. However, the checkbox is not responsive, i.e: clicking it will check itself but it doesn't do anything to the checkboxes below it. 
Expected: Checking off the checkbox should enable all countries (i.e: checkoff all checkboxes associated with countries)

aaayyuusshh committed:
[WIP]: Added a checkbox beside ENABLED

Checking off this checkbox will not enable all countries(check off checkboxes below it) like it is supposed to yet.

KarinG commented:
👍 It’s a checkbox! Great start.

KarinG commented:
This is a Drupal 7 screenshot -> 
![image](https://user-images.githubusercontent.com/5340555/126517257-ba41bc58-c1b5-466a-92bf-d0adb15ee2d2.png)

Note the class `select-all-civi-options`

All JS from D7 including the bit that pertains to the select-all functionality -> 
https://github.com/colemanw/webform_civicrm/blob/7.x-5.x/js/webform_civicrm_options.js#L59-L67

was moved into the D8 (which is also the D9) branch -> it lives here -> https://github.com/colemanw/webform_civicrm/blob/8.x-5.x/js/webform_civicrm_options.js#L59-L67 

So the puzzle now is why is it not working?

KarinG commented:
PS you filed an issue on drupal.org for this. It's good habit to put the link to the drupal issue in the very top of this PR -> hit edit on your description at the top here and add the URL.

aaayyuusshh commented:
Hey Dr.Karin, I have set up php storm, it is ready to  go on my end. I will
edit my issue on drupal ASAP when I get to work and I will also continue
trying to fix the issue. Thank you!

Best regards,
Aayush

On Wed, Jul 21, 2021 at 9:46 AM KarinG ***@***.***> wrote:

> PS you filed an issue on drupal.org for this. It's good habit to put the
> link to the drupal issue in the very top of this PR -> hit edit on your
> description at the top here and add the URL.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/colemanw/webform_civicrm/pull/587#issuecomment-884293564>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ATI3GHN2A7NDLUNYKPZQIWDTY3TUFANCNFSM5AXE5D6A>
> .
>

aaayyuusshh committed:
Merge branch 'colemanw:8.x-5.x' into 8.x-5.x

aaayyuusshh committed:
Merge branch 'colemanw:8.x-5.x' into 8.x-5.x

aaayyuusshh commented:
@KarinG @jitendrapurohit 
Upon adding '#attributes' => ['class' => ['select-all-civi-options']] into 'enabled', the checkbox is still not reactive:
![Screen Shot 2021-08-10 at 6 42 02 PM](https://user-images.githubusercontent.com/80851741/128952626-69b0aba7-623b-4e1d-9983-d7f5014f8484.png)

When I inspect the checkbox, the class is correctly being updated to select-all-civi-options.
This is the inspect tab:
<img width="1438" alt="Screen Shot 2021-08-10 at 6 41 21 PM" src="https://user-images.githubusercontent.com/80851741/128952610-fd1c988b-0a74-45dd-b65b-62c4473cddab.png">

However, unlike the 7th version, it says "select-all-civi-options form-checkbox" rather than "select-all-civi-options wf-civi-processed". I'm wondering what the problem is?
Thank you.

aaayyuusshh committed:
Merge branch 'colemanw:8.x-5.x' into 8.x-5.x

KarinG commented:
@aaayyuusshh @Aashirbadd and I discussed this this morning -> ToDo:

**1. update the PR to include:** ->
```
diff --git a/src/Element/CivicrmSelectOptions.php b/src/Element/CivicrmSelectOptions.php
index 557f0dd..8170420 100644
--- a/src/Element/CivicrmSelectOptions.php
+++ b/src/Element/CivicrmSelectOptions.php
@@ -60,16 +60,16 @@ class CivicrmSelectOptions extends FormElement {
 
     $element['options'] = [
       '#type' => 'table',
-      '#tableselect' => FALSE,
+      '#tableselect' => TRUE,
```

**2. Comment out the javascript in:** https://github.com/colemanw/webform_civicrm/blob/8.x-5.x/js/webform_civicrm_options.js#L59-L67

**3. Add a comment above this section -> that we're looking at using `#tableselect`** - so we're keeping the code for now until we're positive we can remove it

**4. Hide (i.e. `display:none`) the first column of checkboxes** as users need to hit the checkboxes in the enabled column if they wish to select some specific countries

And:
**5. Add a test to ensure the select all is working**

KarinG commented:
With `#tableselect' => TRUE,` -> the select all/deselect all is working nicely:

<img width="839" alt="image" src="https://user-images.githubusercontent.com/5340555/129428092-916bc949-daa7-40a8-bbdc-757faadfd93d.png">

aaayyuusshh committed:
Update '#tableSelect' to TRUE

aaayyuusshh committed:
Merge branch 'colemanw:8.x-5.x' into 8.x-5.x

aaayyuusshh committed:
Update webform_civicrm_options.js

aaayyuusshh commented:
- updated CivicrmSelectOptions.php ✅ : Added #tableSelect => TRUE , since we are trying to use #tableSelect
- updated webform_civicrm_options.js ✅  : Commented out L59-67  (code relating 'select-all-civi-options') since we want to keep the code until we are 100% sure we can remove it.

KarinG commented on js/webform_civicrm_options.js:
```diff
@@ -56,6 +56,10 @@ if (typeof jQuery.fn.prop !== 'function') {
 
+     //Currently we are looking at using #tableSelect thus, we are commenting out 'select-all-civi-options' code for now
+     //until we are certain that we can remove it.
+      
```
@aaayyuusshh - indent this one more space and also put a space between so //Current -> // Current and //until -> // until

KarinG commented:
I see the test is failing -> not sure yet why though.

```
1) Drupal\Tests\webform_civicrm\FunctionalJavascript\CustomFieldSubmissionTest::testSubmitWebform
Behat\Mink\Exception\DriverException: Impossible to check the element with XPath "(//html/.//*
[self::input[not(translate(./@type, 'ABCDEFGHIJKLMNOPQRSTUVWXYZ', 'abcdefghijklmnopqrstuvwxyz') = 'radio' or translate(./@type, 'ABCDEFGHIJKLMNOPQRSTUVWXYZ', 'abcdefghijklmnopqrstuvwxyz')
```

KarinG commented:
Ah - I think I know!

KarinG commented:
In CustomFieldSubmissionTest.php -> we have a test that involves setting the options to 'static' (so that too will now have the select all checkbox).

https://github.com/colemanw/webform_civicrm/blob/8.x-5.x/tests/src/FunctionalJavascript/CustomFieldSubmissionTest.php#L319-320

The error occurs in:
`/home/runner/work/webform_civicrm/webform_civicrm/tests/src/FunctionalJavascript/CustomFieldSubmissionTest.php:361`

Which is this line - so the test can not check a field 'Yes' - there likely is no such field
`$this->getSession()->getPage()->checkField('Yes');`

The thing to do is to recreate the test on an actual Drupal site and to follow its steps.

KarinG commented:
Check this out -> I downloaded the Artifacts -> 

<img width="1383" alt="image" src="https://user-images.githubusercontent.com/5340555/129459744-88eeb621-97aa-44b3-93ba-738c24ceb194.png">

Then opened the archive and looked at the latest screenshot generated for the CustomFieldSubmissionTest -> 
<img width="1151" alt="image" src="https://user-images.githubusercontent.com/5340555/129459751-49c62723-ad8d-48b1-8d77-611a364e84d3.png">

Notice that the radio button has 'Yes' and 'Yes' as options. So I agree with Github Actions - it's impossible to check the 'Yes' option :-)

Question is - how does using `#tableselect` -> cause this?

KarinG commented:
Digging in a little further -> I indeed see the `#tableselect` also applies to the live options (we only want it for static) -> 

<img width="624" alt="image" src="https://user-images.githubusercontent.com/5340555/129459993-d6cd6137-399d-4350-9538-05287f970b72.png">

KarinG commented:
Check out -> https://github.com/colemanw/webform_civicrm/blob/8.x-5.x/src/Element/CivicrmSelectOptions.php#L102-L104

Here we say if we're using live options -> people should not be allowed to drag elements (up and down to resort them). Try adding a line in this if statement  that says:

`$element['options']['#tableselect'] = FALSE;`

KarinG commented:
Yeah - I see this when we're configuring the Custom Radio Field -> somehow this is where we end up with two Yes options.

<img width="644" alt="image" src="https://user-images.githubusercontent.com/5340555/129460227-d699bcb1-4c1e-4a18-85c4-9ff25f50f5df.png">

Will look further in a bit - need to make coffee first!

KarinG commented on js/webform_civicrm_options.js:
```diff
@@ -64,7 +68,8 @@ if (typeof jQuery.fn.prop !== 'function') {
         $('input.civicrm-enabled').change();
-      });
+      }); 
+      */
```
There is an extra space behind the ; here -> 
<img width="309" alt="image" src="https://user-images.githubusercontent.com/5340555/129463644-5f1e6552-262e-4aa5-b2b6-5a42b7459f89.png">

KarinG replied:
👍

KarinG commented on js/webform_civicrm_options.js:
```diff
@@ -56,6 +56,10 @@ if (typeof jQuery.fn.prop !== 'function') {
         }
       }).change();
 
+     //Currently we are looking at using #tableSelect thus, we are commenting out 'select-all-civi-options' code for now
```
There are extra spaces here in the line below:

<img width="357" alt="image" src="https://user-images.githubusercontent.com/5340555/129463673-0d1394b4-04b7-4d22-98e1-9a2b99a40c86.png">

KarinG commented:
I've reproduced the issue:

1. simple cg radio button
2. add to webform
3. switch to static
4. then save

-> note the extra No :-(

<img width="1008" alt="image" src="https://user-images.githubusercontent.com/5340555/129463784-916ac6ca-73e6-41aa-b12e-75c90457dd05.png">

KarinG commented:
PS -> edit element still looks fine -> 
<img width="794" alt="image" src="https://user-images.githubusercontent.com/5340555/129463794-cb4b4703-8767-42b0-b636-55f9dd20a63d.png">

KarinG commented:
So we're doubling up with both [static]checkboxes as well as [static]radios when using `#tableselect`; static[select] works well.

jitendrapurohit commented:
@KarinG @aaayyuusshh looks like the duplicate options are because the code uses the default value callback method to save the options. Adding our custom value_callback to the option list fixes the problem - eg -

```
diff --git a/src/Element/CivicrmSelectOptions.php b/src/Element/CivicrmSelectOptions.php
index 557f0dd..742430d 100644
--- a/src/Element/CivicrmSelectOptions.php
+++ b/src/Element/CivicrmSelectOptions.php
@@ -97,6 +97,7 @@ class CivicrmSelectOptions extends FormElement {
           'group' => 'weight',
         ],
       ],
+      '#value_callback' => [get_called_class(), 'valueCallback'],
     ];
 
     if ($element['#civicrm_live_options']) {
```

aaayyuusshh committed:
Merge branch 'colemanw:8.x-5.x' into 8.x-5.x

aaayyuusshh committed:
Fixed Spacing

aaayyuusshh committed:
Making #tableSelect not apply to Live Options

KarinG commented:
@aaayyuusshh -> don't forget to apply this one line mentioned here -> that will make the test ✅
https://github.com/colemanw/webform_civicrm/pull/587#issuecomment-899035238

aaayyuusshh commented:
@KarinG @jitendrapurohit  Just came across a bug (when #tableSelect=> TRUE in CivicrmSelectOptions.php)

- When all checkboxes are checked off in "Static Options" (using the top right checkbox), you cannot successfully switch back to "Live Options"
![Screen Shot 2021-08-16 at 12 52 05 PM](https://user-images.githubusercontent.com/80851741/129616760-2b0a9955-4488-4a2d-82b9-52888ca0d75d.png)
![Screen Shot 2021-08-16 at 12 52 19 PM](https://user-images.githubusercontent.com/80851741/129615228-7592b40b-b4c3-4cc9-a82b-6d716b82fda0.png)

- When all checkboxes are unchecked (using the top right checkbox), now you can finally successfully switch back to "Live Options" 
![Screen Shot 2021-08-16 at 1 04 12 PM](https://user-images.githubusercontent.com/80851741/129616498-643ede0f-a26d-48b7-980c-1707399963ef.png)
![Screen Shot 2021-08-16 at 1 01 10 PM](https://user-images.githubusercontent.com/80851741/129616505-a718a340-af67-47c6-bff5-c08d9b155235.png)


- Also found out that if you just uncheck some checkboxes manually in Static Options, you still won't be able to successfully switch to Live Options

KarinG commented:
> @KarinG @jitendrapurohit Just came across a bug (when #tableSelect=> TRUE in CivicrmSelectOptions.php)
> 
>     * When all checkboxes are checked off in "Static Options" (using the top right checkbox), you cannot successfully switch back to "Live Options"
>       ![Screen Shot 2021-08-16 at 12 52 05 PM](https://user-images.githubusercontent.com/80851741/129616760-2b0a9955-4488-4a2d-82b9-52888ca0d75d.png)
>       ![Screen Shot 2021-08-16 at 12 52 19 PM](https://user-images.githubusercontent.com/80851741/129615228-7592b40b-b4c3-4cc9-a82b-6d716b82fda0.png)
> 
>     * When all checkboxes are unchecked (using the top right checkbox), now you can finally successfully switch back to "Live Options"
>       ![Screen Shot 2021-08-16 at 1 04 12 PM](https://user-images.githubusercontent.com/80851741/129616498-643ede0f-a26d-48b7-980c-1707399963ef.png)
>       ![Screen Shot 2021-08-16 at 1 01 10 PM](https://user-images.githubusercontent.com/80851741/129616505-a718a340-af67-47c6-bff5-c08d9b155235.png)
> 
>     * Also found out that if you just uncheck some checkboxes manually in Static Options, you still won't be able to successfully switch to Live Options

I don't see this on my local - and the error message re: `File upload 8MB` suggests that something else is causing it. I think we don't need to worry about this at this time. But 👍 for detailing and noting this.

aaayyuusshh committed:
Adding custom value_callback to the option list

KarinG commented:
I need to update my local CiviCRM version [which could break things] so am going to add my notes here on how to get the CSS to work so that it will look like this:

<img width="543" alt="image" src="https://user-images.githubusercontent.com/5340555/129622846-d1e9771d-7eef-4a9c-b0eb-eb39bd6a1672.png">

1. define a new library called 'civicrmoptions' in webform_civicrm.libaries.yml - and add css
2. create that css file in the webform_civicrm/css directory and I think this works (is specific enough)->

```
/* Edit CiviCRM Options Element */
#webform-civicrm-options-wrapper .table-select .form-item {
  display:none;
}
```

3. finally attach the CSS to this edit element form -> in CivicrmSelectOptions.php -> add this line just before we return $element

    $element['#attached']['library'][] = 'webform_civicrm/civicrmoptions';
    return $element;
  }

Ok - now I can go and mess up my local env!

aaayyuusshh commented:
@KarinG I just made all those changes you mentioned above in my local and it looks like the screenshot you attached above.
![Screen Shot 2021-08-16 at 2 28 08 PM](https://user-images.githubusercontent.com/80851741/129625608-f51e1471-2491-4b4e-8f5e-37f077518455.png)


Do you want me to update the PR with these changes as well?

KarinG commented:
Awesome 🎉 and yes - please add those changes to this PR as well!

aaayyuusshh committed:
Added new library: civicrmoptions

aaayyuusshh committed:
Created CSS file that edits options element

aaayyuusshh committed:
Attached CSS
